source/hash_table.c

```diff
@@ -0,0 +1,582 @@
+
+    return AWS_ERROR_SUCCESS;
+}
+
+int aws_common_hash_table_init(struct aws_common_hash_table *map,
+    struct aws_allocator *alloc, size_t size,
+    aws_hash_fn_t hash_fn, aws_equals_fn_t equals_fn, aws_hash_element_destroy_t destroy_fn) {
+
+    struct hash_table_state template;
+    template.hash_fn = hash_fn;
+    template.equals_fn = equals_fn;
+    template.destroy_fn = destroy_fn;
+    template.alloc = alloc;
+
+    template.entry_count = 0;
+    template.max_load_factor = 0.95; // TODO - make configurable?
+
+    update_template_size(&template, size);
+    map->pImpl = alloc_state(&template);
+
+    if (!map->pImpl) {
+        return AWS_ERROR_OOM;
+    }
+
+    return AWS_ERROR_SUCCESS;
+}
+
+void aws_common_hash_table_clean_up(struct aws_common_hash_table *map) {
+    aws_common_hash_table_clear(map);
+    aws_mem_release(((struct hash_table_state *)map->pImpl)->alloc, map->pImpl);
+}
+
+// Tries to find where the requested key is or where it should go if put.
+// Returns AWS_ERROR_SUCCESS if the item existed (leaving it in *entry),
+// or AWS_ERROR_HASHTBL_ITEM_NOT_FOUND if it did not (putting its destination
+// in *entry). Note that this does not take care of displacing whatever was in that entry before.
+
+// probe_idx is set to the probe index of the entry.
+
+static int find_entry1(struct hash_table_state *state, uint64_t hash_code, const void *key, struct hash_table_entry **p_entry, int *p_probe_idx);
+
+// Inlined fast path: Check the first slot, only.
+static int find_entry(struct hash_table_state *state, uint64_t hash_code, const void *key, struct hash_table_entry **p_entry, int *p_probe_idx) __attribute__((always_inline));
+static int find_entry(struct hash_table_state *state, uint64_t hash_code, const void *key, struct hash_table_entry **p_entry, int *p_probe_idx) {
+    struct hash_table_entry *entry = &state->slots[hash_code & state->mask];
+
+    if (entry->hash_code == 0) {
+        if (p_probe_idx) *p_probe_idx = 0;
+        *p_entry = entry;
+        return AWS_ERROR_HASHTBL_ITEM_NOT_FOUND;
+    }
+
+    if (entry->hash_code == hash_code && state->equals_fn(key, entry->element.key)) {
+        if (p_probe_idx) *p_probe_idx = 0;
+        *p_entry = entry;
+        return AWS_ERROR_SUCCESS;
+    }
+
+    return find_entry1(state, hash_code, key, p_entry, p_probe_idx);
+}
+
+static int find_entry1(struct hash_table_state *state, uint64_t hash_code, const void *key, struct hash_table_entry **p_entry, int *p_probe_idx) {
+    int probe_idx = 1;
+    // If we find a deleted entry, we record that index and return it as our probe index (i.e. we'll keep searching to see if it already exists,
+    // but if not we'll overwrite the deleted entry).
+    int probe_cap = INT_MAX;
+
+    int rv;
+    struct hash_table_entry *entry;
+    do {
+        int index = (hash_code + probe_idx) & state->mask;
+        entry = &state->slots[index];
+        if (!entry->hash_code) {
+            rv = AWS_ERROR_HASHTBL_ITEM_NOT_FOUND;
```

Comment:
rv = aws_raise_error(AWS_ERROR_HASHTBL_ITEM_NOT_FOUND) ?

Alternatively, do this, and have a branch at the end that echecks if rv is 0 or -1.

Reply:
same comment for each branch in this loop.

Also it may make since to use the AWS_LIKELY/UNLIKELY in this inner loop.